[DRAFT] Process stealing (Issue #30)

LeedsJohn opened the pull request:
Hi!  This is my work to [implement process stealing](https://github.com/leostera/riot/issues/30).  This is not ready yet!  However, I am feeling a little bit stuck and wanted to ask for advice.

I implemented `steal_processes`, but I was unsure what the `max` parameter was supposed to represent and where it should be passed.  Is it the maximum number of processes a scheduler can steal?

In order to update the `sid` for a process, I had to make that field mutable in `Process.t`.

In the `run` function, I call `steal_processes` instead of sleeping when the queue is empty.

Finally, I have included my work to write a test but it is incomplete.  My understanding is that if I create two processes using the same scheduler (with the new `spawn_pinned` function, one of the processes should be stolen.  However, I am struggling to figure out how to make sure that the process is stolen (and that it's `sid` changes.

I would appreciate any feedback you have!

Thanks!

leostera commented on riot/runtime/scheduler/scheduler.ml:
```diff
@@ -12,6 +12,7 @@ type t = {
   timers : Timer_wheel.t;
   idle_mutex : Mutex.t;
   idle_condition : Condition.t;
+  mutable currently_stealing : bool;
```
Let's make this an `bool Atomic.t` so we guarantee no two schedulers mutate this at the same time.

leostera commented on riot/runtime/scheduler/scheduler.ml:
```diff
@@ -333,6 +335,20 @@ module Scheduler = struct
 
   let tick_timers _pool (sch : t) = Timer_wheel.tick sch.timers
 
+  let steal_processes pool sch =
```
This looks great! Super straight forward. Makes me think that `currently_stealing` could actually be a mutex, so we can wrap this up in a `Mutex.protect` call:

```
Mutex.protect @@ fun () ->
  List.iter (fun scheduler ->
    if Mutex.try_lock scheduler.currently_stealing then
       (* do stealing *)
       Mutex.unlock scheduler.currently_stealing 
  ) pool.schedulers)
```

Something like this, so this guarantees that if A is stealing from B, then A won't get stolen from C, and noone else would be able to steal from B either.

LeedsJohn replied:
That makes sense!  With this, we are guaranteed that a process will only be stolen by one scheduler a time, so it is OK to leave the `sid` in `Process.t` as mutable rather than an `Atomic.t`, right?  (No need to respond if this is correct.)

leostera commented on test/process_stealing_test.ml:
```diff
@@ -0,0 +1,33 @@
+  let this = self () in
+
+  let _ =
+    spawn_pinned (fun () ->
```
We could try having the first process do something more aggressive, like a `while true do .. done` that only yields every 1000 iterations or so.

leostera commented on test/process_stealing_test.ml:
```diff
@@ -0,0 +1,33 @@
+  (* I think that here, the low priority process should be stolen by another
+     scheduler (and it's SID will not equal "00"). However, I'm not sure how to
+     check for this (or if it's working). *)
+  ()
```
For the exit condition, I think we should be able to get process information (like the scheduler id) with a function, so we can compare for both process and it should be different.

I get a feeling that this will only happen as long as:
* A lives longer than B
* A is more resource intensive, so that it hogs its scheduler

What do you think?

leostera commented:
Thanks for opening the PR @LeedsJohn ✨ 

Left you some comments. It overall looks really good, love how straightforward the implementation is!

As soon as we figure out a test, I'd be happy to merge 🚀

LeedsJohn commented:
Hi!  Thank you for the feedback.  I updated I changed `currently_stealing` to a `Mutex.t` and I think that my test is working because the two processes are both created using `spawn_pinned` but eventually have different scheduler IDs.

However, I think that I have broken some stuff and am not sure how to investigate it.  A lot of tests (`link_processes_test`, `process_registration_test`, `cancel_timer_test`, and several others) say `Command got signal SEGV.`  Several others also appear to be broken - for example, `task_test` errors with `Fatal error: exception Stdlib.Effect.Continuation_already_resumed`.

Let me know if you have any thoughts.  Thanks!

leostera commented:
Oh so sorry i forgot to mention that you shouldn't rebase yet, the current main branch is in an amazing/awful place at the moment.

LeedsJohn commented:
No problem, that was my bad - I wasn't sure what the git etiquette was : )

I just reverted my merge, but a lot of the tests seem to still be failing (many with the same `Command got signal SEGV` message).

leostera commented:
I think having the merge commit in there is making Github go bonkers. Could you drop that commit from your history and force-push your branch? Current `main` seems stable again and I've checked that tests are doing ok – other than some Syscall_timeouts that need some attention, everything else should be passing on macOS/Linux.

In general, I'd prefer id you don't merge main back into your branch, but do a rebase instead. That'll help keep the diff view small, and make sure you're always working on top of the latest main.

LeedsJohn commented:
Fixed - thanks for letting me know!

leostera commented:
Hi @LeedsJohn! I've rebased your branch and found some things we need to fix in main before we can merge this. I'll keep you posted 🙏🏼

leostera commented:
Took some refactoring but made it work and boy does it fly! Thanks so much for this @LeedsJohn ✨ 

The gist of the changes I made are:
* move the mutex into the queue, to avoid blocking a scheduler while its being stolen from
* add an explicit check to avoid stealing from ourselves
* refactored the test so it was less flaky on many-core environments (it seldom passed on my 64-core machine)

Most of the other changes are just there to help me debug things, like adding process counts and other metrics to the `Stats` server.

Will be merging now :shipit: – thanks again! 🙏🏼
